x/sisu/chains/eth/bridge.go

```diff
@@ -323,3 +336,115 @@ func (b *bridge) ParseIncomingTx(ctx sdk.Context, chain string, serialized []byt
 func (b *bridge) ProcessCommand(ctx sdk.Context, cmd *types.Command) (*types.TxOut, error) {
 	return nil, fmt.Errorf("Invalid command")
 }
+
+func (b *bridge) ValidateTxOut(ctx sdk.Context, txOut *types.TxOut, transfers []*types.TransferDetails) error {
+	// 1. Validate gas cost.
+	gasInfo, err := b.deyesClient.GetGasInfo(b.chain)
+	if err != nil {
+		return err
+	}
+
+	chainCfg := b.keeper.GetChain(ctx, b.chain)
+	ethCfg := chainCfg.EthConfig
+
+	currentGasCost, _, _ := b.getGasCost(gasInfo, ethCfg.UseEip_1559, gasUnitPerSwap)
+
+	tx := &ethtypes.Transaction{}
+	if err := tx.UnmarshalBinary(txOut.Content.OutBytes); err != nil {
+		return err
+	}
+
+	txGasInfo := &deyesethtypes.GasInfo{}
+	if ethCfg.UseEip_1559 {
+		txGasInfo.Tip = tx.GasTipCap().Int64()
+		txGasInfo.BaseFee = tx.GasFeeCap().Int64()
+	} else {
+		txGasInfo.GasPrice = tx.GasPrice().Int64()
+	}
+
+	txGasCost, _, _ := b.getGasCost(txGasInfo, ethCfg.UseEip_1559, gasUnitPerSwap)
+	if percentage := helper.Difference(txGasCost, currentGasCost); percentage > 3.00 {
+		return fmt.Errorf(
+			"cannot accept the transaction with too large difference in gas cost, diff=%d%%",
+			int(percentage*100))
+	}
+
+	// 2. Validate native token price.
+	currentNativeTokenPrice, err := b.deyesClient.GetTokenPrice(chainCfg.NativeToken)
+	if err != nil {
+		return err
+	}
+
+	txNativeTokenPrice, ok := new(big.Int).SetString(txOut.Input.NativeTokenPrice, 10)
+	if !ok {
+		return fmt.Errorf("cannot convert nativeTokenPrice (%s) to big int", txOut.Input.NativeTokenPrice)
+	}
+
+	if percentage := helper.Difference(currentNativeTokenPrice, txNativeTokenPrice); percentage > 0.1 {
```

Comment:
Using the diff here is incorrect. This is (a -b) /b. What you want is a/b or b/a. 

You should take the absolute ratio a/b and make sure that it's > 0.9 and < 1.1

Reply:
This could be written as 

func CheckRatioThreshold(a, b *big.Int, threshold float32) bool
If b = 0, return false. 
R = a /b 
return R > 1 - threshold && R < 1 + threshold.